langpuppet/gapi.go

```diff
@@ -0,0 +1,327 @@
+	errwrap "github.com/pkg/errors"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/gapi"
+	"github.com/purpleidea/mgmt/pgraph"
+
+	"github.com/purpleidea/mgmt/lang"
+	"github.com/purpleidea/mgmt/puppet"
+
+	"github.com/urfave/cli"
+)
+
+const (
+	// Name is the name of this frontend.
+	Name       = "langpuppet"
+	FlagPrefix = "lp-"
+)
+
+func init() {
+	gapi.Register(Name, func() gapi.GAPI { return &GAPI{} }) // register
+}
+
+// GAPI implements the main langpuppet GAPI interface.
+type GAPI struct {
+	InputURI string
+
+	// should this be a lang.GAPI?
+	langGAPI   gapi.GAPI
+	puppetGAPI gapi.GAPI
+
+	currentLangGraph   *pgraph.Graph
+	currentPuppetGraph *pgraph.Graph
+
+	langGraphReady   bool
+	puppetGraphReady bool
+	graphFlagMutex   *sync.Mutex
+
+	data        gapi.Data
+	initialized bool
+	closeChan   chan struct{}
+	wg          sync.WaitGroup // sync group for tunnel go routines
+}
+
+// Cli takes a cli.Context, and returns our GAPI if activated. All arguments
+// should take the prefix of the registered name. On activation, if there are
+// any validation problems, you should return an error. If this was not
+// activated, then you should return a nil GAPI and a nil error.
+func (obj *GAPI) Cli(c *cli.Context, fs engine.Fs) (*gapi.Deploy, error) {
+	if !c.IsSet(FlagPrefix+lang.Name) && !c.IsSet(FlagPrefix+puppet.Name) {
+		return nil, nil
+	}
+
+	if !c.IsSet(FlagPrefix+lang.Name) || c.String(FlagPrefix+lang.Name) == "" {
+		return nil, fmt.Errorf("%s input is empty", FlagPrefix+lang.Name)
+	}
+	if !c.IsSet(FlagPrefix+puppet.Name) || c.String(FlagPrefix+puppet.Name) == "" {
+		return nil, fmt.Errorf("%s input is empty", FlagPrefix+puppet.Name)
+	}
+
+	flagSet := flag.NewFlagSet(Name, flag.ContinueOnError)
+
+	for _, flag := range c.FlagNames() {
+		if !c.IsSet(flag) {
+			continue
+		}
+		childFlagName := strings.TrimPrefix(flag, FlagPrefix)
+		flagSet.String(childFlagName, "", "no usage string needed here")
+		flagSet.Set(childFlagName, c.String(flag))
+	}
+
+	var langDeploy *gapi.Deploy
+	var puppetDeploy *gapi.Deploy
+	var err error
+
+	// we don't really need the deploy object from the child GAPIs
+	if langDeploy, err = (&lang.GAPI{}).Cli(cli.NewContext(c.App, flagSet, nil), fs); err != nil {
+		return nil, err
+	}
+	if puppetDeploy, err = (&puppet.GAPI{}).Cli(cli.NewContext(c.App, flagSet, nil), fs); err != nil {
+		return nil, err
+	}
+
+	return &gapi.Deploy{
+		Name: Name,
+		Noop: c.GlobalBool("noop"),
+		Sema: c.GlobalInt("sema"),
+		GAPI: &GAPI{
+			InputURI:   fs.URI(),
+			langGAPI:   langDeploy.GAPI,
+			puppetGAPI: puppetDeploy.GAPI,
+		},
+	}, nil
+}
+
+// CliFlags returns a list of flags used by this deploy subcommand.
+// It consists of all flags accepted by lang and puppet mode,
+// with a respective "lp-" prefix.
+func (obj *GAPI) CliFlags() []cli.Flag {
+	langFlags := (&lang.GAPI{}).CliFlags()
+	puppetFlags := (&puppet.GAPI{}).CliFlags()
+
+	var childFlags []cli.Flag
+
```

Comment:
nit: you can nuke this unnecessary nl (i try to add them for readability or logical breaks)